File: plugins/devkit/devkit-java-tests/testData/inspections/cancellationCheckInLoops/NestedLoops.java
package inspections.cancellationCheckInLoops;

import com.intellij.util.concurrency.annotations.RequiresReadLock;

import static inspections.cancellationCheckInLoops.Foo.doSomething;

class Clazz {

  @RequiresReadLock
  public static void foo() {
    int j = 0;
    String[] items = {""};

    // nested loops of different kinds
    while (j < 100) {
      for (int i = 0; i < 5; i++) {
        for (String item : items) {
          <warning descr="Cancellation check 'com.intellij.openapi.progress.ProgressManager.checkCanceled' should be the first statement in a loop body">do</warning> {
            doSomething();
            j++;
          } while (j < 5);
        }
      }
      j++;
    }

    // single line nested loops
    for (int i = 0; i < 5; i++) <warning descr="Cancellation check 'com.intellij.openapi.progress.ProgressManager.checkCanceled' should be the first statement in a loop body">for</warning> (String item : items) {
      doSomething();
    }
  }
}

File: plugins/devkit/devkit-kotlin-tests/testData/inspections/cancellationCheckInLoops/NestedLoops.kt
package inspections.cancellationCheckInLoops

import com.intellij.util.concurrency.annotations.RequiresReadLock

import inspections.cancellationCheckInLoops.Foo.doSomething


@RequiresReadLock
fun main() {
  var j = 0
  // nested loops of different kinds
  for (i in 1..10) {
    while (j < 5) {
      <warning descr="Cancellation check 'com.intellij.openapi.progress.ProgressManager.checkCanceled' should be the first statement in a loop body">do</warning> {
        doSomething()
        j++
      } while (j < 3)
      j++
    }
  }

  // single line nested loops
  for (i in 1..10) <warning descr="Cancellation check 'com.intellij.openapi.progress.ProgressManager.checkCanceled' should be the first statement in a loop body">while</warning> (j < 5) {
    doSomething()
    j++
  }

}